compiler-core/src/type_.rs

```diff
@@ -46,25 +46,71 @@ pub trait HasType {
 
 #[derive(Debug, Clone, PartialEq)]
 pub enum Type {
+    /// App is the primary "wrapper" for most types. The type may come from a
+    /// third party module, but if `module` is empty that means the type is
+    /// coming from the gleam prelude instead (where Int, Float, String, etc is
+    /// defined). If the `App` takes generic arguments they will be put in the
+    /// `args` field.
+    ///
+    /// # Example
+    ///
+    /// ```gleam
+    /// Int //=> Type::App { name: "Int", args: vec![], module: vec![], public: true }
+    /// String //=> Type::App { name: "String", args: vec![], module: vec![], public: true }
+    /// ```
+    ///
     App {
         public: bool,
         module: Vec<String>,
         name: String,
         args: Vec<Arc<Type>>,
     },
 
+    /// Fn defines the type for an _anonymous_ lambda function. These should
+    /// not be confused with module functions which would instead be an
+    /// `Statement::Fn` with its own fields for its arguments and return type
+    ///
+    /// # Example
+    ///
+    /// ```gleam
+    /// fn(x: Int, y: Int) -> Int { x + y } //=> Type::Fn {
+    /// //   args: vec![
+    /// //     Type::App { name: "Int", args: vec![], module: vec![], public: true },
+    /// //     Type::App { name: "Int", args: vec![], module: vec![], public: true },
+    /// //   ],
+    /// //   retrn: Type::App { name: "Int", args: vec![], module: vec![], public: true },
+    /// // }
+    /// ```
+    ///
```

Comment:
I think it's confusing to make the comparison to the function values here, they are unrelated. It's also nothing to do with anonymous functions specifically, that's purely a matter of the value level language. All functions whether named or anonymous has this function type.

```suggestion
    /// The type of a function. It takes arguments and returns a value.
    ///
```